src/parse/text/simple_extensions/types.rs

```diff
@@ -0,0 +1,1773 @@
+                // order.
+                //
+                // So while it might be surprising in some ways that we are not
+                // preserving the order of fields as specified in the YAML, the
+                // nature of YAML somewhat precludes that.
+                //
+                // To give an example: we are considering these two equivalent:
+                //
+                // ```yaml
+                // types:
+                //   - name: point1
+                //     structure:
+                //       longitude: i32
+                //       latitude: i32
+                //   - name: point2
+                //     structure:
+                //       latitude: i32
+                //       longitude: i32
+                // ```
+                //
+                // In Rust, these come in as a `serde_json::Map`, which does not
+                // preserve insertion order. Here, we chose to store keys in
+                // lexicographic order, with an explicit sort so that our
+                // internal representation and any round-trip output remain
+                // deterministic.
+                let mut entries: Vec<_> = field_map.into_iter().collect();
+                entries.sort_by(|a, b| a.0.cmp(&b.0));
+
+                let mut field_names = Vec::new();
+                let mut field_types = Vec::new();
+
+                for (field_name, field_type_value) in entries {
+                    field_names.push(field_name);
+
+                    let type_string = match field_type_value {
+                        serde_json::Value::String(s) => s,
+                        _ => {
+                            return Err(ExtensionTypeError::InvalidFieldType(
+                                "Struct field types must be strings".to_string(),
+                            ));
+                        }
+                    };
+
+                    let parsed_field_type = TypeExpr::parse(&type_string)?;
+                    let mut link = |name: &str| ctx.linked(name);
+                    parsed_field_type.visit_references(&mut link);
+                    let field_concrete_type = ConcreteType::try_from(parsed_field_type)?;
+
+                    field_types.push(field_concrete_type);
+                }
+
+                Ok(ConcreteType {
+                    kind: ConcreteTypeKind::NamedStruct {
+                        field_names,
+                        field_types,
+                    },
+                    nullable: false,
+                })
+            }
+        }
+    }
+}
+
+/// Invalid type name error
+#[derive(Debug, Error, PartialEq)]
+#[error("invalid type name `{0}`")]
+pub struct InvalidTypeName(String);
+
+/// Known Substrait types (builtin + extension references).
+///
+/// Note that this is a recursive type - other than the [BuiltinType]s, the other variants can
+/// have type parameters that are themselves [ConcreteType]s.
```

Comment:
It might be worth including a line of documentation to make the distinction between `Type` and `Kind` here since both words are semantically quite similar. 
```suggestion
/// The structural kind of a Substrait type (builtin, list, map, etc).
///
/// This is almost a complete type, but is missing nullability information. It must be
/// wrapped in a [`ConcreteType`] to form a complete type with nullable/non-nullable annotation.
///
/// Note that this is a recursive type - other than the [BuiltinType]s, the other variants can
/// have type parameters that are themselves [ConcreteType]s.
```